src/execution_plan/ops/op_project.c

```diff
@@ -52,6 +52,9 @@ static Record ProjectConsume(OpBase *opBase) {
 	}
 
 	Record projection = OpBase_CreateRecord(opBase);
+	// Track the inherited Record and the newly-allocated Record so that they may be freed if execution fails.
+	OpBase_AddVolatileRecord(opBase, r);
+	OpBase_AddVolatileRecord(opBase, projection);
```

Comment:
I think if it would be better if we hold both `r` and `projection` within op_project as 2 additional fields,
As `op_project` is the only one using `OpBase_AddVolatileRecord`, `OpBase_RemoveVolatileRecords`
and we call these on each consume.

Reply:
Currently, `OpBase_AddVolatileRecord` and its free are also used in `OpUpdate`. This interface was introduced in #613, at which time it was used in all ops that call `AR_EXP_Evaluate` and have the potential to cause run-time errors.

That probably should still be the case, unless someone can come up with a better way of handling memory leaks on run-time errors.